lint/rule_template_datasource.go

```diff
@@ -2,47 +2,95 @@ package lint
+func (d *Dashboard) checkPrometheusOrLokiDS(template *Template) Result {
+
+	if template.Name != "datasource" {
+		return Result{
+			Severity: Error,
+			Message:  fmt.Sprintf("Dashboard '%s' templated datasource variable named '%s', should be named 'datasource'", d.Title, template.Name),
+		}
+	}
+
+	if template.Label != "Data Source" {
+		return Result{
+			Severity: Error,
+			Message:  fmt.Sprintf("Dashboard '%s' templated datasource variable labeled '%s', should be labeled 'Data Source'", d.Title, template.Label),
+		}
+	}
+
+	if template.Query != Prometheus && template.Query != Loki {
+		return Result{
+			Severity: Error,
+			Message:  fmt.Sprintf("Dashboard '%s' templated datasource variable query is '%s', should be 'prometheus' or 'loki'", d.Title, template.Query),
+		}
+	}
+	return ResultSuccess
+}
+
+func (d *Dashboard) checkPrometheusAndLokiDS(datasources []Template) Result {
+
+	// move loki to 0 index
+	sort.SliceStable(datasources, func(i, j int) bool {
+		return datasources[i].Query < datasources[j].Query
+	})
+
+	if !(datasources[0].Query == Loki && datasources[1].Query == Prometheus) {
+		return Result{
+			Severity: Error,
+			Message:  fmt.Sprintf("Dashboard '%s' with 2 templated datasources should have 'prometheus' and 'loki' types", d.Title),
+		}
+	}
+
+	for _, template := range datasources {
+		prefix := template.Query
+
+		expected_name := prefix + "_datasource"
+		if template.Name != expected_name {
+			return Result{
+				Severity: Error,
+				Message:  fmt.Sprintf("Dashboard '%s' templated datasource variable named '%s', should be named '%s'", d.Title, template.Name, expected_name),
+			}
+		}
+
+		expected_label := strings.Title(prefix) + " Data Source"
+		if template.Label != expected_label {
+			return Result{
+				Severity: Error,
+				Message:  fmt.Sprintf("Dashboard '%s' templated datasource variable labeled '%s', should be labeled '%s'", d.Title, template.Label, expected_label),
+			}
+		}
+	}
+
+	return ResultSuccess
+}
+
 func NewTemplateDatasourceRule() *DashboardRuleFunc {
 	return &DashboardRuleFunc{
 		name:        "template-datasource-rule",
 		description: "Checks that the dashboard has a templated datasource.",
 		fn: func(d Dashboard) Result {
-			template := getTemplateDatasource(d)
-			if template == nil {
-				return Result{
-					Severity: Error,
-					Message:  fmt.Sprintf("Dashboard '%s' does not have a templated datasource", d.Title),
-				}
-			}
-
-			if template.Name != "datasource" {
-				return Result{
-					Severity: Error,
-					Message:  fmt.Sprintf("Dashboard '%s' templated datasource variable named '%s', should be names 'datasource'", d.Title, template.Name),
-				}
-			}
 
-			if template.Label != "Data Source" {
+			datasources := getTemplateDatasources(d)
+			if len(datasources) == 1 {
+				return d.checkPrometheusOrLokiDS(&datasources[0])
+			} else if len(datasources) == 2 {
+				return d.checkPrometheusAndLokiDS(datasources)
+			} else {
```

Comment:
This could be a switch, but also, per the other comments, we only care if there is one datasource template, or more than one.

You could probably combine both of these functions into a single `checkDatasourceTemplates` or similar, which branches depending on `len(...) > 1`.